stdlib/uri.gr

```diff
@@ -0,0 +1,1189 @@
+/**
+ * Utilities for working with URIs.
+ *
+ * @example include "uri"
+ *
+ * @since v0.6.0
+ */
+
+module Uri
+
+include "string"
+include "char"
+include "uint8"
+include "number"
+include "bytes"
+include "buffer"
+include "list"
+include "array"
+include "map"
+include "option"
+include "result"
+
+/**
+ * Represents a parsed RFC 3986 URI.
+ */
+abstract record Uri {
+  scheme: Option<String>,
+  userinfo: Option<String>,
+  host: Option<String>,
+  port: Option<Number>,
+  path: String,
+  query: Option<String>,
+  fragment: Option<String>,
+}
+
+/**
+ * Represents an error encountered while parsing a URI.
+ */
+provide enum ParseError {
+  ParseError,
+}
+
+/**
+ * Represents an error encountered while constructing a URI with `make` or `update`.
+ */
+provide enum ConstructUriError {
+  UserinfoWithNoHost,
+  PortWithNoHost,
+  InvalidSchemeError,
+  InvalidUserinfoError,
+  InvalidHostError,
+  InvalidPortError,
+  InvalidPathError,
+  InvalidQueryError,
+  InvalidFragmentError,
+}
+
+/**
+ * Represents an error encountered while attempting to resolve a URI reference to a target URI.
+ */
+provide enum ResolveReferenceError {
+  BaseNotAbsolute,
+}
+
+/**
+ * Represents an error encountered while attempting to percent-decode a string.
+ */
+provide enum PercentDecodingError {
+  InvalidPercentEncoding,
+}
+
+/**
+ * Used to specify which characters to percent-encode from a string.
+ */
+provide enum PercentEncodeSet {
+  EncodeNonUnreserved,
+  EncodeUserinfo,
+  EncodeRegisteredHost,
+  EncodePath,
+  EncodePathSegment,
+  EncodeQueryOrFragment,
+  EncodeCustom(Char => Bool),
+}
+
+let isDigit = char => {
```

Comment:
Thoughts on moving these into the char lib? This can be changed as well to instead use the operators from the `char` lib:
```
from Char use { (>=), (<=) }
char >= 0 && char <= 9
```